### PR TITLE
[Bugfix] Zero-fill rematerialized weights so reload preserves padding

### PR DESCRIPTION
**Pull request by RyanClark2k**
### Purpose

A same-weights `reload_weights` silently changes any model whose quant method pads weights for kernel alignment. Found while validating #50521 on H100: gpt-oss-20b greedy output diverged from the first token after reloading the identical checkpoint, on stock main, with all storage addresses preserved.

Root cause: quant methods that round weight dims up for kernel alignment (for mxfp4, `mxfp4_round_up_hidden_size_and_intermediate_size` pads gpt-oss's 2880 to 2944 on every backend) create those weights zero-initialized, and weight loaders only ever write the checkpoint-backed slices, so correctness of the pad regions depends on the initial zeros. Layerwise reload restores the construction-time parameter set on meta and rematerializes it with `torch.empty_strided`, which is uninitialized. After streaming refills the real slices, the pad regions hold allocator garbage, the layer under-loads (`load_numel < load_numel_total`) into the delayed-processing path, and `process_weights_after_loading` packs the garbage into the kernel weights. Garbage pad rows of `w13_weight` and pad entries of `w13_bias` produce nonzero activations at pad intermediate channels, and garbage pad columns of `w2_weight` contract those into every real hidden output, so all logits change.

The fix zero-fills the tensor in `materialize_meta_tensor`. This restores the `create_weights` initial state for every region a loader does not cover, making reload byte-identical to first load. Cost is one memset per restored tensor during reload, which is negligible against the weight copy itself.

This maps to RFC #48312 category 4 (reload state preservation) and the category 3 padding-accounting subtype, whose text already flags padding as needing explicit treatment.

### Notes for reviewers

- Zero is the correct fill for the padded-weights pattern because `create_weights` zero-initializes these tensors. A parameter whose initial value is non-zero and is only partially loaded would still be wrong after this fix, but it is deterministic-wrong instead of garbage-wrong, and no such site is known; the fail-closed direction (declaring loadable regions explicitly) is part of the #48478 discussion.
- The diagnosis that led here: md5 over every parameter and buffer localized the drift to exactly the 138 padded MoE tensors, with all other model parameters byte-identical and zero storage moves. The exact-zero count dropped from 475,537,003 to 273,111,316 after one reload, and per layer `w13_bias` lost exactly 32 experts x 128 pad elements of zeros. Hashes changed again on a second reload (87/96), ruling out a deterministic double-transform and fingerprinting uninitialized memory.

### Test Plan

Unit tests in `tests/model_executor/model_loader/test_reload.py`:

- `test_materialize_meta_tensor_zero_fills`: the direct contract. Poisons the allocator with freed NaN-filled blocks so unfixed code demonstrably returns garbage, then asserts materialization yields zeros.
- `test_layerwise_reload_preserves_padded_weight_zeros`: end-to-end layerwise cycle with a padded layer whose loader writes only the leading columns. The layer under-loads, takes the delayed-processing path at finalization, and the test asserts the kernel storage address is preserved, the loaded slice carries the new values, and the pad region comes back as zeros.

Both discriminate: with `zero_()` removed, both fail under allocator poisoning.

H100 NVL validation on gpt-oss-20b (mxfp4, flashinfer_cutlass backend, CUDA graphs enabled), vLLM wheel at the base commit with only `reload/meta.py` overlaid for green. Oracle: md5 over all 96 MoE weights, scales, and biases across two same-weights reloads, plus greedy token comparison.

- Red (stock): 138/144 tensor hashes change after one same-weights reload, and 135/144 change again between reload 1 and reload 2 (non-deterministic, fingerprinting uninitialized memory). Zero elements drop from 475,537,003 to 311,188,846. Greedy tokens diverge from the first position.
- Green (this fix): 0/144 hash changes across two reloads and the zero count is byte-identical. Reload reproduces first-load bytes exactly.
- Combined with #50521: greedy tokens match baseline after both reload 1 and reload 2. With only one of the two fixes applied, tokens still diverge (this fix alone leaves the graph-captured constants storage freed and reclaimed by the reload's own allocations; #50521 alone leaves the padded weights corrupted), so the pair jointly restores end-to-end same-weights reload correctness for this model, and each PR's own oracle isolates its half.

### Test Result

- Both new tests pass; the pre-existing reload unit tests pass; ruff lint and format clean.
- H100 red/green: see above.

Related: RFC #48312, #48478, #50521 (whose H100 validation surfaced this defect and documented the contaminated token oracle).

This PR includes AI-generated code (Claude Code). I reviewed every changed line, and validated the fix end-to-end on H100 hardware with the red/green protocol described above.

### CODE REVIEW

**Comment by RyanClark2k** [comment]
Attaching the diagnosis and red/green harness with full results for the H100 validation described in the test plan.

Environment: H100 NVL (SM90), vLLM wheel built at commit 4689c7dd6 (the reload machinery is unchanged between that commit and this PR's base) with only `reload/meta.py` overlaid for the green run, flashinfer-python 0.6.15.post1, gpt-oss-20b, `moe_backend="flashinfer_cutlass"`, CUDA graphs enabled. 24 MoE layers times 6 padded tensors (w13/w2 weights, scales, biases) = 144 hashed tensors.

| Metric | red (stock main) | green (this PR) |
|---|---|---|
| Tensor hashes changed, first load to reload 1 | 138/144 | 0/144 |
| Tensor hashes changed, reload 1 to reload 2 | 135/144 | 0/144 |
| Exact-zero elements, first load to reload 1 | 475,537,003 to 311,188,846 | unchanged |
| Greedy tokens match baseline after reload | no | no with this fix alone, yes combined with #50521 |

The reload-to-reload hash drift on red (135/144) is the fingerprint of uninitialized memory: a deterministic double-transform defect would produce identical bytes on every reload. The zero-element deltas match the padding arithmetic exactly, for example `w13_bias` has 4096 zeros at first load, which is 32 experts times 128 pad elements (2 x (2944 - 2880)), and loses most of them after reload on red.

On the token oracle: gpt-oss-20b on this path has two independent reload defects, this padding corruption and the graph-captured SwiGLU constants rebind fixed by #50521. With only this fix, weights are byte-identical but the reload's own allocations reclaim the freed capture-time constants storage, so tokens still diverge. With both fixes overlaid, greedy tokens match baseline after reload 1 and reload 2. Each PR's own oracle (byte hashes here, pointer census there) isolates its half; the combined run shows the pair restores end-to-end same-weights reload correctness for this model.

<details>
<summary>Raw results</summary>

```json
{
  "label": "red",
  "tensors": 144,
  "hash_changed_load_to_reload1": 138,
  "hash_changed_reload1_to_reload2": 135,
  "zeros_first_load": 475537003,
  "zeros_after_reload1": 311188846,
  "tokens_match_reload1": false,
  "tokens_match_reload2": false,
  "changed_keys_sample": [
    "model.layers.0.mlp.experts.routed_experts.w13_bias",
    "model.layers.0.mlp.experts.routed_experts.w13_weight",
    "model.layers.0.mlp.experts.routed_experts.w13_weight_scale",
    "model.layers.0.mlp.experts.routed_experts.w2_bias",
    "model.layers.0.mlp.experts.routed_experts.w2_weight",
    "model.layers.0.mlp.experts.routed_experts.w2_weight_scale"
  ]
}
```

```json
{
  "label": "green",
  "tensors": 144,
  "hash_changed_load_to_reload1": 0,
  "hash_changed_reload1_to_reload2": 0,
  "zeros_first_load": 475537003,
  "zeros_after_reload1": 475537003,
  "tokens_match_reload1": false,
  "tokens_match_reload2": false,
  "changed_keys_sample": []
}
```

Combined with #50521:

```json
{
  "label": "both",
  "tensors": 144,
  "hash_changed_load_to_reload1": 0,
  "hash_changed_reload1_to_reload2": 0,
  "zeros_first_load": 475537003,
  "zeros_after_reload1": 475537003,
  "tokens_match_reload1": true,
  "tokens_match_reload2": true,
  "changed_keys_sample": []
}
```

</details>

<details>
<summary>Red/green harness</summary>

Run as `python validate_padding_fix.py --label red|green|both` with `VLLM_ALLOW_INSECURE_SERIALIZATION=1`.

```python
# SPDX-License-Identifier: Apache-2.0
"""H100 red/green validation for the zero-fill materialization fix.

Oracle: byte-exact md5 over every gpt-oss-20b MoE weight, scale, and bias
across same-weights reloads, plus greedy token comparison. Red (stock)
corrupts the padded tensors on every reload non-deterministically; green
(zero-fill fix) must reproduce the first-load bytes exactly and keep greedy
output identical.

Usage: python validate_padding_fix.py --label red|green
"""

import argparse
import json

MODEL = "openai/gpt-oss-20b"
PROMPTS = [
    "The capital of France is",
    "In a shocking turn of events, the",
    "def fibonacci(n):",
    "The three laws of thermodynamics are",
]


def moe_hashes(model):
    import hashlib

    import torch

    out = {}
    for name, module in model.named_modules():
        if not name.endswith("routed_experts"):
            continue
        for pname in (
            "w13_weight",
            "w2_weight",
            "w13_weight_scale",
            "w2_weight_scale",
            "w13_bias",
            "w2_bias",
        ):
            t = getattr(module, pname, None)
            if t is None or t.is_meta:
                continue
            raw = t.detach().contiguous().cpu().reshape(-1).view(torch.uint8)
            out[f"{name}.{pname}"] = hashlib.md5(raw.numpy().tobytes()).hexdigest()
    return out


def zero_count(model):
    out = 0
    for name, module in model.named_modules():
        if not name.endswith("routed_experts"):
            continue
        for pname in ("w13_weight", "w2_weight", "w13_bias", "w2_bias"):
            t = getattr(module, pname, None)
            if t is None or t.is_meta:
                continue
            out += int((t.detach().float() == 0).sum().item())
    return out


def generate(llm):
    from vllm import SamplingParams

    params = SamplingParams(temperature=0.0, max_tokens=32)
    outputs = llm.generate(PROMPTS, params)
    return [list(o.outputs[0].token_ids) for o in outputs]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--label", required=True, choices=["red", "green", "both"])
    args = parser.parse_args()
    label = args.label

    from vllm import LLM

    llm = LLM(
        model=MODEL,
        moe_backend="flashinfer_cutlass",
        max_model_len=2048,
        enable_prefix_caching=False,
        gpu_memory_utilization=0.85,
    )

    hashes0 = llm.apply_model(moe_hashes)[0]
    zeros0 = llm.apply_model(zero_count)[0]
    tokens0 = generate(llm)
    print(f"[{label}] first load: {len(hashes0)} tensors, {zeros0} zero elements")

    llm.collective_rpc("reload_weights", kwargs={"weights_path": MODEL})
    hashes1 = llm.apply_model(moe_hashes)[0]
    zeros1 = llm.apply_model(zero_count)[0]
    tokens1 = generate(llm)

    llm.collective_rpc("reload_weights", kwargs={"weights_path": MODEL})
    hashes2 = llm.apply_model(moe_hashes)[0]
    tokens2 = generate(llm)

    diff01 = sorted(k for k in hashes0 if hashes0[k] != hashes1[k])
    diff12 = sorted(k for k in hashes1 if hashes1[k] != hashes2[k])

    total = len(hashes0)
    print(f"\n===== RESULT [{label}] =====")
    print(f"tensors hashed:                      {total}")
    print(f"hashes changed, load -> reload 1:    {len(diff01)}/{total}")
    print(f"hashes changed, reload 1 -> 2:       {len(diff12)}/{total}")
    print(f"zero elements, load -> reload 1:     {zeros0} -> {zeros1}")
    print(f"tokens match after reload 1:         {tokens1 == tokens0}")
    print(f"tokens match after reload 2:         {tokens2 == tokens0}")

    with open(f"padfix_result_{label}.json", "w") as f:
        json.dump(
            {
                "label": label,
                "tensors": total,
                "hash_changed_load_to_reload1": len(diff01),
                "hash_changed_reload1_to_reload2": len(diff12),
                "zeros_first_load": zeros0,
                "zeros_after_reload1": zeros1,
                "tokens_match_reload1": tokens1 == tokens0,
                "tokens_match_reload2": tokens2 == tokens0,
                "changed_keys_sample": diff01[:6],
            },
            f,
            indent=2,
        )
    print(f"wrote padfix_result_{label}.json")


if __name__ == "__main__":
    main()
```

</details>

<details>
<summary>Diagnosis probe that established the mechanism</summary>

```python
# SPDX-License-Identifier: Apache-2.0
"""Confirm the uninitialized-padding mechanism for gpt-oss mxfp4 reload drift.

Probe 1: exact-zero counts and abs-max of the MoE biases and weights before
and after reload. create_weights zero-fills padded checkpoint-format params;
reload materializes them with empty_strided, so pad regions should show
garbage (fewer zeros, larger abs-max) if the hypothesis holds.

Probe 2: reload twice. Uninitialized garbage differs per materialization, so
the kernel-weight hashes should differ between reload 1 and reload 2. A
deterministic double-transform defect would produce identical hashes instead.

Usage: python diag2_padding_probe.py
"""

import json

MODEL = "openai/gpt-oss-20b"


def moe_stats(model):
    import torch

    out = {}
    for name, module in model.named_modules():
        if not name.endswith("routed_experts"):
            continue
        for pname in ("w13_bias", "w2_bias", "w13_weight", "w2_weight"):
            t = getattr(module, pname, None)
            if t is None or t.is_meta:
                continue
            tf = t.detach().float()
            out[f"{name}.{pname}"] = {
                "zeros": int((tf == 0).sum().item()),
                "numel": t.numel(),
                "absmax": float(tf.abs().max().item()),
                "nan_or_inf": int((~torch.isfinite(tf)).sum().item()),
            }
    return out


def moe_hashes(model):
    import hashlib

    import torch

    out = {}
    for name, module in model.named_modules():
        if not name.endswith("routed_experts"):
            continue
        for pname in ("w13_weight", "w2_weight", "w13_weight_scale", "w13_bias"):
            t = getattr(module, pname, None)
            if t is None or t.is_meta:
                continue
            raw = t.detach().contiguous().cpu().reshape(-1).view(torch.uint8)
            out[f"{name}.{pname}"] = hashlib.md5(raw.numpy().tobytes()).hexdigest()
    return out


def summarize(label, stats):
    total_zeros = sum(v["zeros"] for v in stats.values())
    worst_absmax = max(v["absmax"] for v in stats.values())
    nonfinite = sum(v["nan_or_inf"] for v in stats.values())
    print(f"[{label}] total exact zeros: {total_zeros}, "
          f"worst absmax: {worst_absmax:.4g}, nonfinite: {nonfinite}")
    sample = sorted(stats)[0]
    print(f"[{label}] sample {sample}: {stats[sample]}")
    return {"total_zeros": total_zeros, "worst_absmax": worst_absmax,
            "nonfinite": nonfinite}


def main():
    from vllm import LLM

    llm = LLM(
        model=MODEL,
        moe_backend="flashinfer_cutlass",
        max_model_len=2048,
        enable_prefix_caching=False,
        gpu_memory_utilization=0.85,
    )

    stats0 = llm.apply_model(moe_stats)[0]
    hashes0 = llm.apply_model(moe_hashes)[0]
    s0 = summarize("first load", stats0)

    llm.collective_rpc("reload_weights", kwargs={"weights_path": MODEL})
    stats1 = llm.apply_model(moe_stats)[0]
    hashes1 = llm.apply_model(moe_hashes)[0]
    s1 = summarize("reload 1", stats1)

    llm.collective_rpc("reload_weights", kwargs={"weights_path": MODEL})
    stats2 = llm.apply_model(moe_stats)[0]
    hashes2 = llm.apply_model(moe_hashes)[0]
    s2 = summarize("reload 2", stats2)

    diff01 = sorted(k for k in hashes0 if hashes0[k] != hashes1[k])
    diff12 = sorted(k for k in hashes1 if hashes1[k] != hashes2[k])

    print("\n===== PADDING PROBE RESULT =====")
    print(f"tensors hashed: {len(hashes0)}")
    print(f"hash diffs first load -> reload 1: {len(diff01)}")
    print(f"hash diffs reload 1 -> reload 2:   {len(diff12)}")
    if diff12:
        print("reload-to-reload drift confirms NON-DETERMINISTIC source "
              "(uninitialized memory), not a deterministic double-transform")
    zeros_delta = s0["total_zeros"] - s1["total_zeros"]
    print(f"exact-zero count first load vs reload 1: "
          f"{s0['total_zeros']} -> {s1['total_zeros']} (delta {zeros_delta})")
    print(f"bias/weight absmax: {s0['worst_absmax']:.4g} -> "
          f"{s1['worst_absmax']:.4g}; nonfinite {s0['nonfinite']} -> "
          f"{s1['nonfinite']}")

    with open("diag2_result.json", "w") as f:
        json.dump(
            {
                "summary": {"first_load": s0, "reload1": s1, "reload2": s2},
                "hash_diff_load_vs_reload1": len(diff01),
                "hash_diff_reload1_vs_reload2": len(diff12),
                "sample_stats_first": {k: stats0[k] for k in sorted(stats0)[:4]},
                "sample_stats_reload1": {k: stats1[k] for k in sorted(stats1)[:4]},
            },
            f,
            indent=2,
        )
    print("wrote diag2_result.json")


if __name__ == "__main__":
    main()
```

</details>
